src/dataflow/src/source/s3.rs

```diff
@@ -267,15 +292,27 @@ async fn scan_bucket_task(
                 continuation_token = response.next_continuation_token;
             }
             Err(e) => {
+                let err_string = format!("{}", e);
+                tx.send(Err(S3Error::ListObjectsFailed(e)))
+                    .await
+                    .unwrap_or_else(|e| {
+                        log::debug!("unable to send error on listing objects: {}", e)
+                    });
+
                 allowed_errors -= 1;
                 if allowed_errors == 0 {
-                    log::error!("failed to list bucket {}: {}", bucket, e);
+                    log::error!("failed to list bucket {}: {}", bucket, err_string);
```

Comment:
I am able to get this error  to happen by setting toxiproxy to terminate TCP streams that are longer than a certain number of bytes:

```
2021-06-09T08:50:42.930804Z  WARN dataflow::source::s3: unable to list bucket testdrive-test-1: Error obtaining body: Error obtaining chunk: error reading a body from connection: end of file before message length reached (1 retries remaining)    
2021-06-09T08:50:43.951284Z ERROR dataflow::source::s3: failed to list bucket testdrive-test-1: Error obtaining body: Error obtaining chunk: error reading a body from connection: end of file before message length reached 
```

however, it does not cause the source to error out:

```materialize=> CREATE MATERIALIZED SOURCE s1
  FROM S3
  DISCOVER OBJECTS
  MATCHING 'short/*'
  USING BUCKET SCAN 'testdrive-test-1',
  SQS NOTIFICATIONS 'testdrive-test-1'
  WITH (
    region = 'us-east1',
    endpoint = 'http://toxiproxy:4566',
    access_key_id = 'a',
    secret_access_key = 'b',
    token = 'c'
  )
  FORMAT TEXT;
CREATE SOURCE
materialize=> select * from s1;
 text | mz_record 
------+-----------
(0 rows)
```